lib/AST/ASTScope.cpp

```diff
@@ -76,6 +76,34 @@ void ASTScope::dumpOneScopeMapLocation(std::pair<unsigned, unsigned> lineCol) {
   impl->dumpOneScopeMapLocation(lineCol);
 }
 
+bool ASTScope::isLocWithinMacroDeclOrTopLevelMacroExpansionDeclScope(
+    SourceFile *sf, SourceLoc loc) {
+  // The innermost scope containing the loc.
+  auto *innermostScope = sf->getScope().impl->findInnermostEnclosingScope(
+      loc, nullptr);
+  auto isMDOrTopLevelMED = [&](const ASTScopeImpl *scope) -> bool {
+    if (auto *decl = scope->getDeclIfAny().getPtrOrNull()) {
+      if (auto *med = dyn_cast<MacroExpansionDecl>(decl)) {
+        if (med->getDeclContext()->getContextKind() ==
+            DeclContextKind::FileUnit)
+          return true;
+      } else if (auto *md = dyn_cast<MacroDecl>(decl)) {
```

Comment:
Why check for a macro declaration here? It's only macro expansions that we need to worry about, right?

Reply:
When we populate `SourceLookupCache` with top-level macro expansion results, type-checking a `MacroDecl`'s arguments will run into the same circular dependency problem as type-checking a top-level `MacroExpansionDecl`'s arguments. So I needed to check for both MD and MED.